src/components/Input/Input.tsx

```diff
@@ -104,120 +110,119 @@ class Input extends AutoControlledComponent<Extendable<IInputProps>, any> {
+    })
+  }
+
+  private handleInputRef = (c: HTMLInputElement) => (this.inputRef = c)
 
-  state: any = { value: this.props.value || this.props.defaultValue || '' }
+  private handleIconOverrides = predefinedProps => ({
+    onClick: (e: React.SyntheticEvent) => {
+      this.handleOnClear()
+      this.inputRef.focus()
+      _.invoke(predefinedProps, 'onClick', e, this.props)
+    },
+    ...(predefinedProps.onClick && { tabIndex: '0' }),
+  })
 
-  handleChange = e => {
+  private handleChange = (e: React.SyntheticEvent) => {
     const value = _.get(e, 'target.value')
 
     _.invoke(this.props, 'onChange', e, { ...this.props, value })
 
     this.trySetState({ value })
   }
 
-  handleChildOverrides = (child, defaultProps) => ({
-    ...defaultProps,
-    ...child.props,
-  })
-
-  handleInputRef = c => (this.inputRef = c)
-
-  handleOnClear = e => {
-    const { clearable } = this.props
-
-    if (clearable) {
+  private handleOnClear = () => {
+    if (this.props.clearable) {
       this.trySetState({ value: '' })
     }
   }
 
-  partitionProps = () => {
-    const { type } = this.props
-    const { value } = this.state
-
-    const unhandled = getUnhandledProps(Input, this.props)
-    const [htmlInputProps, rest] = partitionHTMLProps(unhandled)
-
-    return [
-      {
-        ...htmlInputProps,
-        onChange: this.handleChange,
-        type,
-        value: value || '',
-      },
-      rest,
-    ]
-  }
-
-  computeIcon = () => {
+  private computeIcon = (): ShorthandValue => {
     const { clearable, icon } = this.props
     const { value } = this.state
 
-    if (clearable && value.length !== 0) {
+    if (clearable && (value as string).length !== 0) {
       return 'close'
     }
 
-    if (!_.isNil(icon)) return icon
-
-    return null
-  }
-
-  handleIconOverrides = predefinedProps => {
-    return {
-      onClick: e => {
-        this.handleOnClear(e)
-
-        this.inputRef.focus()
-        _.invoke(predefinedProps, 'onClick', e, this.props)
-      },
-      ...(predefinedProps.onClick && { tabIndex: '0' }),
-    }
-  }
-
-  renderComponent({ ElementType, classes, styles, variables }) {
-    const { renderIcon, renderInput, type } = this.props
-    const [htmlInputProps, restProps] = this.partitionProps()
-
-    const inputClasses = classes.input
-
-    return (
-      <ElementType className={classes.root} {...restProps}>
-        {createHTMLInput(type, {
-          defaultProps: htmlInputProps,
-          overrideProps: {
-            className: inputClasses,
-            ref: this.handleInputRef,
-          },
-          render: renderInput,
-        })}
-        {this.computeIcon() &&
-          Icon.create(this.computeIcon(), {
-            defaultProps: {
-              styles: styles.icon,
-              variables: variables.icon,
-            },
-            overrideProps: this.handleIconOverrides,
-            render: renderIcon,
-          })}
-      </ElementType>
-    )
+    return icon || null
```

Comment:
here we will ignore any overrides provided to icon slot in case we have `clearable` input with non-empty value - this is not what client would expect, especially given that there might be no `close` icon provided by user's theme